jupyter_kernel_test/__init__.py

```diff
@@ -188,11 +219,124 @@ def test_display_data(self):
+            self.assertEqual(len(reply['content']['history'][0]), 3)
+            self.assertEqual(codes, [h[2] for h in reply['content']['history']])
+
+            session, start = reply['content']['history'][0][0:2]
+            with self.subTest(output=True):
+                reply = self.history_helper(codes, output=True, raw=True,
+                                            hist_access_type="tail", n=n)
+                self.assertEqual(len(reply['content']['history'][0][2]), 2)
+
+        with self.subTest(hist_access_type="range"):
+            if 'range' not in self.supported_history_operations:
+                raise SkipTest
+            if session is None:
+                raise SkipTest
+            reply = self.history_helper(codes, output=False, raw=True,
+                                        hist_access_type="range",
+                                        session=session, start=start,
+                                        stop=start+1)
+            self.assertEqual(len(reply['content']['history']), 1)
+            self.assertEqual(reply['content']['history'][0][0], session)
+            self.assertEqual(reply['content']['history'][0][1], start)
+
+        with self.subTest(hist_access_type="search"):
+            if not self.code_history_pattern:
+                raise SkipTest
+            if 'search' not in self.supported_history_operations:
+                raise SkipTest
+
+            with self.subTest(subsearch="normal"):
+                reply = self.history_helper(codes, output=False, raw=True,
+                                            hist_access_type="search",
+                                            pattern=self.code_history_pattern)
+                self.assertGreaterEqual(len(reply['content']['history']), 1)
+            with self.subTest(subsearch="unique"):
+                reply = self.history_helper(codes, output=False, raw=True,
+                                            hist_access_type="search",
+                                            pattern=self.code_history_pattern,
+                                            unique=True)
+                self.assertEqual(len(reply['content']['history']), 1)
+            with self.subTest(subsearch="n"):
+                reply = self.history_helper(codes, output=False, raw=True,
+                                            hist_access_type="search",
+                                            pattern=self.code_history_pattern,
+                                            n=3)
+                self.assertEqual(len(reply['content']['history']), 3)
+
+    code_inspect_sample = ""
+
+    def test_inspect(self):
+        if not self.code_inspect_sample:
+            raise SkipTest
+
+        for i in (0, 1):
+            with self.subTest(detail_level=i):
+                self.flush_channels()
+                msg_id = self.kc.inspect(self.code_inspect_sample,
+                                         detail_level=i)
```

Comment:
detail_level is another thing that I don't know how far different kernels implement. Again, we copied part of the IPython infrastructure into the message protocol, and never really thought about whether it made sense. I'm not even sure if any of our frontends use `detail_level=1`.

Let's just test it with the default level for now. I'd rather start with fewer tests and add them as we determine they're useful than add more and decide to remove things.